cpp/src/arrow/ipc/writer.cc

```diff
@@ -1076,6 +1076,11 @@ class PayloadFileWriter : public internal::IpcPayloadWriter, protected StreamBoo
   }
 
   Status Close() override {
+    // Close stream for compatibility with sequential readers
+    RETURN_NOT_OK(UpdatePosition());
+    int32_t end_of_stream_marker = 0;
+    RETURN_NOT_OK(Write(&end_of_stream_marker, sizeof(int32_t)));
```

Comment:
Can you move the `kEos` constexpr out of the `PayloadStreamWriter` to global scope and use that here? Then you don't need `end_of_stream_marker`